<a href="https://colab.research.google.com/github/JuanJopz08/AnaliticaNeg/blob/main/Reto_2_Arboles_de_decision.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Reto 2

0. libreria de trabajo

In [ ]:
import numpy as np
import pandas as pd

#para implementación y evalución del modelo:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1. base de datos

In [ ]:
nxl = '/content/2. TechSalesReport_Int_L.xlsx'
df = pd.read_excel(nxl,sheet_name=0)

df2 = df[['Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS','College']]
df2['Business'] = df2['Business'].map({'Hardware': 0, 'Software': 1})
df2['College'] = df2['College'].map({'Yes': 1, 'No': 0})

XD=df2[['Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS']]
yd=df2[['College']]

<ipython-input-7-52958a28c359>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Business'] = df2['Business'].map({'Hardware': 0, 'Software': 1})
<ipython-input-7-52958a28c359>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['College'] = df2['College'].map({'Yes': 1, 'No': 0})


2. modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
#Max Depth es el número de variables disponibles
#Gini es el criterio que indica el grado de error en el modelo

mar.fit(XD,yd)

#Se procede con el despliegue del arbol de decisión
from sklearn.tree import export_graphviz #Exportar
from pydotplus import graph_from_dot_data #Crear el gráfico

vs=['Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS']
dot_data=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png('1.TechSalesReport.png')

True

3. Desempeño modelo

In [ ]:
ydp=mar.predict(XD)
cm=confusion_matrix(yd,ydp)
print(cm)


VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

Exactitud=((VP+VN)/(VP+VN+FP+FN)) #Funcionamiento general del modelo
print('La Exactitud del modelo es del',Exactitud)
Sensibilidad=(VP/(VP+FN)) #Como se comporto el modelo con los positivos
print('La Sensibilidad del modelo es del',Sensibilidad)
Especificidad=(VN/(VN+FP)) #Como se comporto el modelo con los 0
print('La Especificidad del modelo es del',Especificidad)
Precision=(VP/(VP+FP)) #Cual es el acierto clasificando los positivos
print('La Precision del modelo es del',Precision)
TasaError=((FP+FN)/(VP+VN+FP+FN)) #Equivocación general
print('La Tasa de Error del modelo es del',TasaError)
PredNeg=(VN/(VN+FN)) #Como funciona el modelo calificando en 0.
print('La PredNeg es del',PredNeg)

[[  114  4356]
 [   69 17451]]
La Exactitud del modelo es del 0.7987721691678036
La Sensibilidad del modelo es del 0.9960616438356165
La Especificidad del modelo es del 0.025503355704697986
La Precision del modelo es del 0.8002476269087908
La Tasa de Error del modelo es del 0.20122783083219645
La PredNeg es del 0.6229508196721312


4. Naive Bayes

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd.values.ravel())

GaussianNB()

5. Pronostico titulo Universitario

In [ ]:
XDB2=pd.read_excel(nxl,sheet_name=1)
XDB2.head(100) #Solo muestra 100 datos

# Convertir Business a valores numericos
XDB2['Business']=pd.Categorical(XDB2['Business']).codes
XDB2=XDB2.dropna()
print(XDB2)

dfp=mnb.predict

    Sales_Rep  Business  Age  Female  Years Personality  Certificates  \
0           1         0   63       0      2    Sentinel             5   
1           2         1   38       1      2    Diplomat             4   
2           3         1   47       0      1     Analyst             2   
3           4         1   25       0      1    Explorer             2   
4           5         0   49       0     10    Explorer             4   
5           6         1   48       0      1    Diplomat             1   
6           7         1   25       1      1    Explorer             6   
7           8         0   39       0      1    Diplomat             2   
8           9         1   29       0      1    Explorer             5   
9          10         1   45       1      3    Diplomat             3   
10         11         0   35       0      6    Diplomat             1   
11         12         0   30       0      4    Diplomat             2   
12         13         1   28       0      1     Ana

6. Numero de personas y nodos

In [ ]:
# Total de personas:
total_personas = len(XDB)

# Número de personas con título universitario (inscritas):
num_inscritas = XDB['College'].sum()

# Número de personas sin título universitario (no inscritas):
num_no_inscritas = total_personas - num_inscritas

# Imprimir los resultados:
print("Total de personas en la base de datos:",total_personas)
print("Número de personas con título universitario (inscritas):",num_inscritas)
print("Número de personas sin título universitario (no inscritas):",num_no_inscritas)

# Determinar el número de nodos puros:
# Obtener el número total de nodos en el árbol:
n_nodes = mar.tree_.node_count

# Obtener la información de las salidas por izquierda y derecha de cada nodo:
children_left = mar.tree_.children_left
children_right = mar.tree_.children_right

# Contar los nodos puros (nodos hoja donde todos los registros pertenecen a la misma clase):
pure_nodes = sum(1 for i in range(n_nodes) if children_left[i] == children_right[i])

# Imprimir el resultado:
print("Número de nodos puros:",pure_nodes)

Total de personas en la base de datos: 21990
Número de personas con título universitario (inscritas): 17520
Número de personas sin título universitario (no inscritas): 4470
Número de nodos puros: 16


7. Excel

In [ ]:
#Se toma la totalidad de los solicitante
XDB2=pd.read_excel (nxl, sheet_name=0)
XDB2.head (100)
yd2=XDB2[['College']]
yd2.head(16)

#Donde estan los que no poseen informacion de PreApr
filas=np.where((np.isnan(yd2))==True)[0]
print('Las filas que no poseen informacion:',filas)

#se hace el pronostico de la PreApr para estas personas
XD2=XDB2[['Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS']]
XD2=XD2.iloc[filas,]

#Se hace el pronostico para estas personas
ydp2=mar.predict(XD2)
print('Pronostico:',ydp2)

#Crear el informe con la PreApr
dfp=pd.DataFrame(np.column_stack((filas,XD2,ydp2)))
dfp.columns=['filas','Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS','College']
dfp.head(16)
dfp.to_excel('InformeCollege.xlsx')

#Cuantos fueron positivos y cuantos negativos
print('los negados pronosticados son:',len(np.where(ydp2==0)[0]))
print('los aprobados pronosticados son:',len(np.where(ydp2==1)[0]))

TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

ANÁLISIS DE RESULTADOS

El modelo fue construido con un total de 21,990 registros, de los cuales 17,520 correspondían a empleados con título universitario y 4,470 sin título. Para su implementación, se consideraron 9 variables en total, incluyendo 8 variables de entrada: Business, Age, Years, Female, Feedback, Salary, Certificates y NPS; y una variable de salida encargada de determinar si un empleado posee un título universitario.

Después de ejecutar el modelo de Árbol de Decisión, se obtuvo una exactitud del 79%, superando el umbral del 75% y evidenciando un desempeño adecuado en la clasificación. Además, el coeficiente de sensibilidad fue cercano al 100%, lo que demuestra su capacidad para identificar correctamente los casos positivos. Por otro lado, la precisión del modelo alcanzó el 80%, lo que confirma su fiabilidad en el pronóstico y su correcto balance en la clasificación de los datos.